Mathlib/NumberTheory/LSeries/Convolution.lean

```diff
@@ -0,0 +1,203 @@
+/-
+Copyright (c) 2024 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import Mathlib.Analysis.InnerProductSpace.Basic
+import Mathlib.NumberTheory.LSeries.Convergence
+import Mathlib.Analysis.Normed.Field.InfiniteSum
+
+/-!
+# Dirichlet convolution of sequences and products of L-series
+
+We define the *Dirichlet convolution* `f ⍟ g` of two sequences `f g : ℕ → R` with values in a
+semiring `R` by `(f ⍟ g) n = ∑ (k * m = n), f k * g m` when `n ≠ 0` and `(f ⍟ g) 0 = 0`.
+Technically, this is done by transporting the existing definition for `ArithmeticFunction R`;
+see `LSeries.convolution`. We show that these definitions agree (`LSeries.convolution_def`).
+
+We then consider the case `R = ℂ` and show that `L (f ⍟ g) = L f * L g` on the common domain
+of convergence of the L-series `L f`  and `L g` of `f` and `g`; see `LSeries_convolution`
+and `LSeries_convolution'`.
+-/
+
+open scoped LSeries.notation
+
+open Complex LSeries
+
+/-!
+### Dirichlet convolution of two functions
+-/
+
+open BigOperators
+
+/-- We turn any function `ℕ → R` into an `ArithmeticFunction R` by setting its value at `0`
+to be zero. -/
+def toArithmeticFunction {R : Type*} [Zero R] (f : ℕ → R) : ArithmeticFunction R where
+  toFun n := if n = 0 then 0 else f n
+  map_zero' := rfl
+
+/-- If we consider an arithmetic function just as a function and turn it back into an
+arithmetic function, it is the same as before. -/
+@[simp]
+lemma ArithmeticFunction.toArithmeticFunction_eq_self {R : Type*} [Zero R]
+    (f : ArithmeticFunction R) :
+    toArithmeticFunction f = f := by
+  ext n
+  simp (config := {contextual := true}) [toArithmeticFunction, ArithmeticFunction.map_zero]
+
+/-- Dirichlet convolution of two sequences.
+
+We define this in terms of the already existing definition for arithmetic functions. -/
+noncomputable def LSeries.convolution {R : Type*} [Semiring R] (f g : ℕ → R) : ℕ → R :=
+  (toArithmeticFunction f * toArithmeticFunction g :)
+
+@[inherit_doc]
+scoped[LSeries.notation] infixl:70 " ⍟ " => LSeries.convolution
+
+open scoped LSeries.notation
+
+/-- The product of two arithmetic functions defines the same function as the Dirichlet convolution
+of the functions defined by them. -/
+lemma ArithmeticFunction.mul_to_convolution {R : Type*} [Semiring R] (f g : ArithmeticFunction R) :
+    f ⍟ g = (f * g :) := by
+  simp only [convolution, ArithmeticFunction.toArithmeticFunction_eq_self]
```

Comment:
The name doesn't seem right; perhaps `ArithmeticFunction.coe_mul`, or even `ArithmeticFunction.coe_mul_eq_convolution`? I would also rather spell the conclusion as `f ⍟ g = ↑(f * g)`.

Reply:
OK. I take `ArithmeticFunction.coe_mul`; the alternative does not fit so well as the equality is written the other way round.